### PR TITLE
Fix for Windows: std.os.windows.DeleteFile()

### PR DESCRIPTION
**Pull request by suirad**
This commit addresses #5537 using the following test:

```zig
const builtin = @import("builtin");
const std = @import("std");
const fs = std.fs;
const assert = std.debug.assert;
const print = std.debug.print;
const testing = std.testing;

test "#5537" {
    if (builtin.os.tag != .windows) return;
    const cwd = fs.cwd();

    cwd.makeDir("tmp") catch |e| {
        switch (e) {
            error.PathAlreadyExists => {},
            else => return e,
        }
    };
    defer cwd.deleteDir("tmp") catch unreachable;

    var file = try cwd.createFile("tmp/tmpfile", .{ .read = true });
    defer {
        file.close();
        cwd.deleteFile("tmp/tmpfile") catch unreachable;
    }

   // Actual test for proper behavior
    testing.expectError(error.DirNotEmpty, cwd.deleteDir("tmp"));
}
```

I considered adding this test to the std to catch regressions but I wasn't sure if it was wanted to be added and I also wasn't sure exactly where in the repo to add it.

### CODE REVIEW

**Comment by squeek502** [comment]
Here's a possible test case that can be added to [`std/fs/test.zig`](https://github.com/ziglang/zig/blob/master/lib/std/fs/test.zig):

```zig
test "deleteDir" {
    var tmp_dir = tmpDir(.{});
    defer tmp_dir.cleanup();

    // deleting a non-existent directory
    testing.expectError(error.FileNotFound, tmp_dir.dir.deleteDir("test_dir"));

    var dir = try tmp_dir.dir.makeOpenPath("test_dir", .{});
    var file = try dir.createFile("test_file", .{});
    file.close();
    dir.close();

    // deleting a non-empty directory
    testing.expectError(error.DirNotEmpty, tmp_dir.dir.deleteDir("test_dir"));

    dir = try tmp_dir.dir.openDir("test_dir", .{});
    try dir.deleteFile("test_file");
    dir.close();

    // deleting an empty directory
    try tmp_dir.dir.deleteDir("test_dir");
}
```

**Comment by kubkon** [comment]
> Nice. Here's a possible test case that can be added to [`std/fs/test.zig`](https://github.com/ziglang/zig/blob/master/lib/std/fs/test.zig):
> 
> ```zig
> test "deleteDir" {
>     var tmp_dir = tmpDir(.{});
>     defer tmp_dir.cleanup();
> 
>     // deleting a non-existent directory
>     testing.expectError(error.FileNotFound, tmp_dir.dir.deleteDir("test_dir"));
> 
>     var dir = try tmp_dir.dir.makeOpenPath("test_dir", .{});
>     var file = try dir.createFile("test_file", .{});
>     file.close();
>     dir.close();
> 
>     // deleting a non-empty directory
>     testing.expectError(error.DirNotEmpty, tmp_dir.dir.deleteDir("test_dir"));
> 
>     dir = try tmp_dir.dir.openDir("test_dir", .{});
>     try dir.deleteFile("test_file");
>     dir.close();
> 
>     // deleting an empty directory
>     try tmp_dir.dir.deleteDir("test_dir");
> }
> ```

We definitely need a test case added for this, thanks for suggesting this @squeek502!

**Comment by kubkon** [comment]
@suirad, have you perhaps experimented with [ReactOS approach](https://github.com/reactos/reactos/blob/893a3c9d030fd8b078cbd747eeefd3f6ce57e560/dll/win32/kernel32/client/file/dir.c#L941) where instead of setting `FILE_DELETE_ON_CLOSE` in `NtCreateFile`, we open the file/dir with `DELETE` and then mark it for deletion with `NtSetInformationFile`. I'm curious to see whether this will have different behaviour when trying to remove a non-empty directory, i.e., ideally yield an error of some kind rather than succeed and ignore the op altoghether.

File: lib/std/os.zig [hunk 1]
```diff
@@ -1887,7 +1887,13 @@ pub fn unlinkatZ(dirfd: fd_t, file_path_c: [*:0]const u8, flags: u32) UnlinkatEr
...
 pub fn unlinkatW(dirfd: fd_t, sub_path_w: []const u16, flags: u32) UnlinkatError!void {
     const remove_dir = (flags & AT_REMOVEDIR) != 0;
-    return windows.DeleteFile(sub_path_w, .{ .dir = dirfd, .remove_dir = remove_dir });
+    try windows.DeleteFile(sub_path_w, .{ .dir = dirfd, .remove_dir = remove_dir });
+    if (remove_dir){
+        faccessatW(dirfd, @ptrCast([*:0]const u16, sub_path_w.ptr), 0, 0) catch 
+            return;
+        return error.DirNotEmpty;
+    }
+    return;
```
Comment (kubkon) [inline lib/std/os.zig:1896]: Do we want to check for this *only* in `std.os.unlinkatW`, or should we actually move this into `std.os.windows.DeleteFile`? A Windows user is not obligated to use `std.os.unlinkatW` after all. In fact, some would probably prefer to use `std.os.windows` as the native Windows API.

At the very least, if this is not the plan, we should add a comment to that effect in `std.os.windows.DeleteFile` documenting the fact that Windows may silently fail to delete a non-empty directory, although, I think I'd personally feel more comfortable with  `DeleteFile` returning an error condition instead.

Comment (kubkon) [reply lib/std/os.zig:1896]: Also, should this not be extended to `std.os.rmdirW`?

**Review by kubkon — CHANGES_REQUESTED** [review]

File: lib/std/os.zig (hunk 1 above)
Comment (suirad) [reply lib/std/os.zig:1896]: Yea it seems you are right that the fix would be better off in `std.os.windows.DeleteFile` to always apply. Though it is a bit tilting that there are quite a few ways to do the same thing that aren't coordinated together(in regards to `std.os.rmdirW` and `std.os.unlinkW` which seem to be the same thing. `rmdir` should at least alias to `unlink` with a `remove_dir` flag.

I may address that inconsistancy as well.

Comment (kubkon) [reply lib/std/os.zig:1896]: Oh, actually, `rmdir` and `unlink` are two separate calls in POSIX (and I should add that `std.os` is ultimately going to be renamed to `std.posix`). [`rmdir`](https://www.man7.org/linux/man-pages/man2/rmdir.2.html) is meant to be used to remove directories, while [`unlink`](https://www.man7.org/linux/man-pages/man2/unlink.2.html) to remove other entities like files and symlinks. It is quite unfortunate that the later added syscall `unlinkat` blurs that line. Therefore, I'd keep them separate from each other on Windows to emphasise that fact. Also, having both call `std.os.windows.DeleteFile` directly shaves off one function call if `rmdirW` was an "alias" to `unlinkW`.

**Comment by suirad** [comment]
I've updated the code adding @squeek502 's test and moving the non empty directory detection into `std.os.windows.DeleteFile` per @kubkon 's notes.

File: lib/std/os/windows.zig [hunk 2]
```diff
@@ -818,14 +819,27 @@ pub fn DeleteFile(sub_path_w: []const u16, options: DeleteFileOptions) DeleteFil
...
+        .SUCCESS => CloseHandle(tmp_handle),
         .OBJECT_NAME_INVALID => unreachable,
         .OBJECT_NAME_NOT_FOUND => return error.FileNotFound,
         .INVALID_PARAMETER => unreachable,
         .FILE_IS_A_DIRECTORY => return error.IsDir,
         .NOT_A_DIRECTORY => return error.NotDir,
         else => return unexpectedStatus(rc),
     }
+
+    if (options.remove_dir){
```
Comment (kubkon) [inline lib/std/os/windows.zig:831]: Would you mind adding a comment for posterity explaining why we need to test for the dir existence here?

**Review by kubkon — APPROVED** [review]
Looks great! Just one more nitpick which doesn't tweak the behaviour of this PR, and we're good to go.

File: lib/std/os/windows.zig (hunk 2 above)
Comment (suirad) [reply lib/std/os/windows.zig]: Added @kubkon
